### PR TITLE
Consider all exceptions as possible execution results

### LINKED ISSUE #170: Handle all exceptions as possible results
However, if unexpected exceptions have been thrown, Lincheck should print their stack traces.

### PR DESCRIPTION
**Pull request by avpotapov00**
Fixes #170

### CODE REVIEW

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/InternalLincheckExceptionEmulator.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/InternalLincheckExceptionEmulator.kt:18]: Should be MPL, please check other files

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/annotations/Operation.java [hunk 2]
```diff
@@ -49,7 +49,10 @@
 
     /**
      * Handle the specified exceptions as a result of this operation invocation.
+     *
+     * @deprecated all exceptions now handled as possible result
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/annotations/Operation.java:53]: add "s." at the end

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt [hunk 3]
```diff
@@ -160,7 +160,7 @@ internal class FixedActiveThreadsExecutor(private val nThreads: Int, runnerHash:
             try {
                 runnable.run()
             } catch(e: Throwable) {
-                setResult(iThread, wrapInvalidAccessFromUnnamedModuleExceptionWithDescription(e))
+                setResult(iThread, wrapExceptionIfNecessary(e))
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/FixedActiveThreadsExecutor.kt:163]: It's unclear what "necessary" means. In fact, there is only one case when Lincheck handles exception differently -- when an error related to Java modules and missed JVM options has occurred. Let's rename the "wrapExceptionIfNecessary" correspondingly, to make its contract explicit.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/TestThreadExecutionGenerator.java [hunk 4]
```diff
@@ -256,7 +265,6 @@ private static void generateRun(ClassVisitor cv, Type testType, int iThread, Lis
             mv.visitLabel(unexpectedExceptionHandler);
             // Call onFailure method
             mv.dup();
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/TestThreadExecutionGenerator.java:267]: You can add a special "failOnExceptionIsUnexpected(e: Throwable)" method that
1) checks whether the exception cannot be a result
2) if everything is fine, finishes 
3) if not, calls Runner.onFailure(..) method and throws the exception.

This way, you need neither this bytecode generation nor the added in this PR one.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 5]
```diff
@@ -49,10 +49,9 @@ internal fun executeActor(
         val res = m.invoke(instance, *args.toTypedArray())
         return if (m.returnType.isAssignableFrom(Void.TYPE)) VoidResult else createLincheckResult(res)
     } catch (invE: Throwable) {
-        val eClass = (invE.cause ?: invE).javaClass.normalize()
-        for (ec in actor.handledExceptions) {
-            if (ec.isAssignableFrom(eClass))
-                return ExceptionResult.create(eClass)
+        val exception = invE.cause ?: invE
+        if (exceptionCanBeValidExecutionResult(exception)) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:53]: Is this statement always "true"?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 6]
```diff
@@ -49,10 +49,9 @@ internal fun executeActor(
         val res = m.invoke(instance, *args.toTypedArray())
         return if (m.returnType.isAssignableFrom(Void.TYPE)) VoidResult else createLincheckResult(res)
     } catch (invE: Throwable) {
-        val eClass = (invE.cause ?: invE).javaClass.normalize()
-        for (ec in actor.handledExceptions) {
-            if (ec.isAssignableFrom(eClass))
-                return ExceptionResult.create(eClass)
+        val exception = invE.cause ?: invE
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:52]: What does this line mean? Please clarify and document,

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/generator/ParamGeneratorsTests.kt [hunk 7]
```diff
@@ -9,14 +9,16 @@
...
+package org.jetbrains.kotlinx.lincheck_test.generator
 
-import junit.framework.Assert.*
+import junit.framework.Assert.assertTrue
 import org.jetbrains.kotlinx.lincheck.*
-import org.jetbrains.kotlinx.lincheck.annotations.*
-import org.jetbrains.kotlinx.lincheck.paramgen.*
-import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.*
-import org.junit.*
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/generator/ParamGeneratorsTests.kt:16]: Please configure your IDE to ALWAYS use imports with "*" for Kotlin code. Update all the changed files correspondingly.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ForcibleFinishExceptionInTryBlockTest.kt [hunk 8]
```diff
@@ -8,15 +8,16 @@
...
 import org.jetbrains.kotlinx.lincheck.*
 import org.jetbrains.kotlinx.lincheck.annotations.*
 import org.jetbrains.kotlinx.lincheck.strategy.*
 import org.jetbrains.kotlinx.lincheck.strategy.managed.*
 import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.*
-import org.jetbrains.kotlinx.lincheck.test.*
+import org.jetbrains.kotlinx.lincheck_test.*
 import org.junit.*
 
 class ForcibleFinishExceptionInTryBlockTest {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ForcibleFinishExceptionInTryBlockTest.kt:23]: What if we add an empty `catch(t: Throwable) {}` block?

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/runner/ParallelThreadsRunnerExceptionTest.kt [hunk 9]
```diff
@@ -152,6 +153,33 @@ class ParallelThreadsRunnerExceptionTest {
     }
 }
 
+class ParallelThreadExecutionExceptionsTest {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/runner/ParallelThreadsRunnerExceptionTest.kt:156]: Please move this test to a separate file and update the test name. You also do not need to mock anything here.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/TransformedExceptionTests.kt [hunk 10]
```diff
@@ -34,7 +35,7 @@ class UnexpectedTransformedExceptionTest : AbstractLincheckTest(UnexpectedExcept
         throwException = true
         throwException = false
         if (throwException)
-            throw CustomException()
+            throw InternalLincheckTestException
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/transformation/TransformedExceptionTests.kt:38]: Let's rename the exception name to show what is so unusual in it.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/InternalLincheckExceptionEmulator.kt [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+@file:Suppress("PackageDirectoryMismatch")
+
+package org.jetbrains.kotlinx.lincheck.util
+
+/**
+ * Utility object to emulate exception in the Lincheck itself
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/util/InternalLincheckExceptionEmulator.kt:16]: Please explain that the package name differs, and that is how Lincheck understands whether the code internal or not, blah-blah-blah.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/verifier/linearizability/ClocksTest.kt [hunk 12]
```diff
@@ -1,4 +1,5 @@
-/*
+/*-
+ * #%L
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/verifier/linearizability/ClocksTest.kt:2]: Revert this change

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionAsResultTest.java [hunk 13]
```diff
@@ -1,4 +1,5 @@
 /*
+ * #%L
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionAsResultTest.java:2]: Revert this change

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionAsResultTest.java [hunk 14]
```diff
@@ -8,7 +9,7 @@
  * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-package org.jetbrains.kotlinx.lincheck.test;
+package org.jetbrains.kotlinx.lincheck_test;
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/ExceptionAsResultTest.java:12]: In some tests, the package has not been changed. Please check all files.

File: src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+Init part:
+[operation2(): void, operation1(): void, operation1(): void, operation2(): void, operation2(): void]
+Parallel part:
+| operation1(): void       | operation2(): IllegalStateException #1 [1,-] |
+| operation1(): void [1,-] | operation1(): void                     [1,1] |
+Post part:
+[operation1(): void, operation1(): void, operation1(): void, operation2(): void, operation1(): void]
+
+---
+values in "[..]" brackets indicate the number of completed operations
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt:11]: Please start this sentence with a capital letter.

File: src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+| operation1(): void [1,-] | operation1(): void                     [1,1] |
+Post part:
+[operation1(): void, operation1(): void, operation1(): void, operation2(): void, operation1(): void]
+
+---
+values in "[..]" brackets indicate the number of completed operations 
+in each of the parallel threads seen at the beginning of the current operation
+---
+Number after exception name is used to match exception with it's 
+stacktrace in the exceptions section below the interleaving section
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt:15]: Let's try to make the message shorter. Suggestion: "The number next to an exception name helps you find its stack trace provided after the interleaving section."

File: src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+|                                                                                                                                | operation2(): IllegalStateException #1 |
+|                                                                                                                                | operation1(): void                     |
+|                                                                                                                                |   thread is finished                   |
+|   canEnterForbiddenSection.WRITE(false) at ClocksWithExceptionsInOutputTest.operation1(ClocksWithExceptionsInOutputTest.kt:25) |                                        |
+|   result: void                                                                                                                 |                                        |
+|   thread is finished                                                                                                           |                                        |
+
+Exceptions traces:
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt:31]: Exception stack traces

File: src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+---
+values in "[..]" brackets indicate the number of completed operations 
+in each of the parallel threads seen at the beginning of the current operation
+---
+Number after exception name is used to match exception with it's 
+stacktrace in the exceptions section below the interleaving section
+---
+
+= The following interleaving leads to the error =
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt:18]: Do we need this title?

File: src/jvm/test/resources/expected_logs/internal_bug_report.txt [hunk 19]
```diff
@@ -0,0 +1,11 @@
+It seems you've found a bug in Lincheck.
+Please report it to JetBrains.
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/internal_bug_report.txt:2]: Why to JetBrains? How?

"Please submit an issue here https://github.com/JetBrains/lincheck/issues. We kindly ask providing a stack trace printed below and the code that causes the error." (please improve this text)

File: src/jvm/test/resources/expected_logs/internal_bug_report.txt [hunk 20]
```diff
@@ -0,0 +1,11 @@
+It seems you've found a bug in Lincheck.
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/internal_bug_report.txt:1]: "Wow! You've catched a bug in Lincheck."

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    @Operation
+    fun operation2() {
+        if (canEnterForbiddenSection) throwException()
+    }
+
+    @Test
+    fun `should add stackTrace to output`() {
+        // We will remove line with TestThreadExecution call as it's number may vary
+        val testRunnableLineRegex = "(\\W*)at org.jetbrains.kotlinx.lincheck.runner.TestThreadExecution(\\d+)\\.run\\(Unknown Source\\)".toRegex()
+        val expectedOutput = """
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:40]: Please specify the expected output in a file, similarly to other tests

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.checkImpl
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingOptions
+import org.jetbrains.kotlinx.lincheck.util.InternalLincheckExceptionEmulator.throwException
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+@Suppress("UNUSED")
+class InternalLincheckBugTest {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:21]: Please explain shortly what this test checks

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ExceptionsInOutputTest.kt [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.jetbrains.kotlinx.lincheck_test.representation
+
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck_test.util.runModelCheckingTestAndCheckOutput
+import org.junit.Test
+
+@Suppress("UNUSED")
+class ExceptionsInOutputTest {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ExceptionsInOutputTest.kt:18]: Please explain shortly what this test checks

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ClocksWithExceptionsInOutputTest.kt [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package org.jetbrains.kotlinx.lincheck_test.representation
+
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck_test.util.runModelCheckingTestAndCheckOutput
+import org.junit.Test
+
+@Suppress("UNUSED")
+class ClocksWithExceptionsInOutputTest {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ClocksWithExceptionsInOutputTest.kt:18]: Please explain shortly what this test checks

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt [hunk 25]
```diff
@@ -60,18 +60,44 @@ object Cancelled : Result() {
...
  */
 @Suppress("DataClassPrivateConstructor")
-data class ExceptionResult private constructor(val tClazz: Class<out Throwable>, override val wasSuspended: Boolean) : Result() {
+class ExceptionResult private constructor(
+    /**
+     * Exception is stored to print it's stackTrace in case of incorrect results
+     */
+    val throwable: Throwable,
+    /**
+     * Normalized version of exception class
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt:69]: of THE exception class.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt [hunk 26]
```diff
@@ -60,18 +60,44 @@ object Cancelled : Result() {
...
+     */
+    val tClazz: Class<out Throwable>,
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt:71]: Do you need it? You can compare ExceptionResult's by class names.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 27]
```diff
@@ -80,28 +80,53 @@ private class ActorWithResult(val actorRepresentation: String, val spacesAfterAc
                                     " ".repeat(spacesAfterResult) + clockRepresentation
 }
 
-private fun uniteActorsAndResultsLinear(actors: List<Actor>, results: List<Result>): List<ActorWithResult> {
+private fun uniteActorsAndResultsLinear(
+    actors: List<Actor>,
+    results: List<Result>,
+    exceptionsDescription: Map<Throwable, ExceptionDescription>
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:86]: exceptionStackTraces?

Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:86]: ExceptionDescription -> ExceptionNumberAndStacktrace

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 28]
```diff
@@ -136,18 +161,29 @@ internal fun StringBuilder.appendExecutionScenario(scenario: ExecutionScenario):
 }
 
 internal fun StringBuilder.appendFailure(failure: LincheckFailure): StringBuilder {
+    val results: ExecutionResult? = (failure as? IncorrectResultsFailure)?.results
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:164]: In Reporter and TraceReporter, please document everything you touch. These files are hard to read, so we need improve them.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 29]
```diff
@@ -226,4 +366,6 @@ private fun StringBuilder.appendException(t: Throwable) {
     val sw = StringWriter()
     t.printStackTrace(PrintWriter(sw))
     appendln(sw.toString())
-}
\ No newline at end of file
+}
+
+private const val LINCHECK_PACKAGE_NAME = "org.jetbrains.kotlinx.lincheck."
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:371]: Move to Utils?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 30]
```diff
@@ -171,44 +274,81 @@ private fun StringBuilder.appendDeadlockWithDumpFailure(failure: DeadlockWithDum
...
     }
-    if (failure.results.parallelResultsWithClock.flatten().any { !it.clockOnStart.empty })
-        appendln("\n---\nvalues in \"[..]\" brackets indicate the number of completed operations \n" +
-            "in each of the parallel threads seen at the beginning of the current operation\n---")
+
+    val hints = mutableListOf<String>()
+    if (failure.results.parallelResultsWithClock.flatten().any { !it.clockOnStart.empty }) {
+        hints.add(
+            """
+                values in "[..]" brackets indicate the number of completed operations
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:328]: According to this description, we need to replace "-" in clocks with "0" for better clarity.

**Review by ndkoval — CHANGES_REQUESTED** [review]
That's a great improvement for Lincheck! From the high level, everything looks good. Let's address several issues I've raised, and the PR should be good to go.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt (hunk 5 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:53]: Under some circumstances some exception can be thrown here, and I think it's good that we will process it in the same way

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ForcibleFinishExceptionInTryBlockTest.kt (hunk 8 above)
Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/ForcibleFinishExceptionInTryBlockTest.kt:23]: We will fail with proper result anyway, but a little bit later.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt (hunk 5 above)
Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:53]: I agree, but which exceptions are invalid here?

File: src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt (hunk 18 above)
Comment (avpotapov00) [reply src/jvm/test/resources/expected_logs/clocks_and_exceptions.txt:18]: I don't see any verbosity here; it's just one line describing the following section.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt (hunk 26 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt:70]: Are you talking about simple class names? If yes, there might be situations with exceptions with the same names from different packages. If it's about full class name, then do you remember why you initially stored normalized version of class, do we transform exceptions?

File: src/jvm/test/resources/expected_logs/internal_bug_report.txt [hunk 31]
```diff
@@ -0,0 +1,11 @@
+Wow! You've caught a bug in Lincheck.
+We kindly ask to provide an issue here https://github.com/JetBrains/lincheck/issues,
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/internal_bug_report.txt:2]: Please add ":" before the link

File: src/jvm/test/resources/expected_logs/illegal_module_access.txt [hunk 32]
```diff
@@ -0,0 +1,38 @@
+= The execution failed with an unexpected exception =
+Execution scenario (parallel part):
+| incrementTwice() | operation() |
+
+java.lang.RuntimeException: It seems that you use Java 9+ and the code uses Unsafe or similar constructions that are not accessible from unnamed modules.
+Please add the following lines to your test running configuration:
+--add-opens java.base/jdk.internal.misc=ALL-UNNAMED
+--add-export   s java.base/jdk.internal.util=ALL-UNNAMED
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/illegal_module_access.txt:8]: whitespace in "exports"

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Operation
+    fun operation2() {
+        if (canEnterForbiddenSection) throwException()
+    }
+
+    @Test
+    fun `should add stackTrace to output`() = runModelCheckingTestAndCheckOutput(
+        expectedOutputFile = "internal_bug_report.txt",
+        linesToRemoveRegex = TEST_EXECUTION_TRACE_ELEMENT_REGEX,
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:42]: Why do you need to remove some lines?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt [hunk 34]
```diff
@@ -49,12 +48,14 @@ internal fun executeActor(
...
-            if (ec.isAssignableFrom(eClass))
-                return ExceptionResult.create(eClass)
-        }
-        throw IllegalStateException("Invalid exception as a result of $actor", invE)
+        /*
+        The actor Method is invoked using Reflection API,
+        so if some exception occurs, it will be wrapped with InvocationTargetException,
+        using its `cause` property.
+        But exception may be thrown not only by method invocation,
+        so in that case we have to store the exception itself.
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt:56]: If the exception is thrown not during the method invocation, shouldn't we fail immediately?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt (hunk 26 above)
Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Result.kt:70]: I mean the full class name. No need to store the `Class` instance.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt [hunk 35]
```diff
@@ -80,28 +80,53 @@ private class ActorWithResult(val actorRepresentation: String, val spacesAfterAc
...
+): List<ActorWithResult> {
     require(actors.size == results.size) {
         "Different numbers of actors and matching results found (${actors.size} != ${results.size})"
     }
     return actors.indices.map {
-        ActorWithResult("${actors[it]}", 1, "${results[it]}", 0, "")
+        val result = results[it]
+        ActorWithResult(
+            actorRepresentation = "${actors[it]}",
+            spacesAfterActor = 1,
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:95]: wouldn't it be better to specify the total length instead? Now, you have an invariant that `actorRepresentation.size + spacesAfterActor == resultRepresentation.size + spacesAfterResult`, which is easy to break accidentally.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt [hunk 36]
```diff
@@ -18,23 +18,44 @@ import kotlin.math.*
...
+    val startTraceGraphNode = constructTraceGraph(scenario, results, trace, exceptionStackTraces)
 
     appendln(PARALLEL_PART_TITLE)
     val traceRepresentation = traceGraphToRepresentationList(startTraceGraphNode, false)
     appendTraceRepresentation(scenario, traceRepresentation)
     appendln()
-
     appendln()
+
+    if (exceptionStackTraces.isNotEmpty()) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/TraceReporter.kt:32]: You append exception stack traces for both stress testing and model checking modes. This information is not related to the interleaving traces. I would, therefore, split this function to "appendShortTrace" and "appendDetailedTrace" functions, calling them separately. This way, you can easily print the stack traces between the traces.

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/HBClock.kt [hunk 37]
```diff
@@ -17,7 +17,7 @@ data class HBClock(val clock: IntArray) {
     operator fun get(i: Int) = clock[i]
 
     override fun toString() = clock.joinToString(prefix = "[", separator = ",", postfix = "]") {
-        if (it == 0) "-" else "$it"
+        if (it == 0) "0" else "$it"
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/HBClock.kt:20]: This logic does nothing now :)

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt (hunk 33 above)
Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:42]: to filter expected and actual output to remove test suite-dependent lines, for example TestThreadExecution number depends on test suite

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt (hunk 35 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:95]: I suggest to make refactoring in a separate ticket.

Comment (ndkoval) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:95]: Sure, let's refactor Reporter and TraceReporter files, but let's also not introduce new design mistakes.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt (hunk 33 above)
Comment (ndkoval) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:42]: Can you mention that in the code?

Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:42]: It's already mentioned in Kdoc to `runModelCheckingTestAndCheckOutput` function.

Comment (ndkoval) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/InternalLincheckBugTest.kt:42]: I see, but add a comment indicating which lines you are removing.

File: src/jvm/test/resources/expected_logs/illegal_module_access.txt (hunk 32 above)
Comment (ndkoval) [reply src/jvm/test/resources/expected_logs/illegal_module_access.txt:8]: still here

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt (hunk 35 above)
Comment (avpotapov00) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/Reporter.kt:95]: No, here I don't have such constrains (`actorRepresentation.size + spacesAfterActor == resultRepresentation.size + spacesAfterResult`) this method is called only to constuct output for Init and Post parts.

**Pull request merged**